zeal/rpc/rpc.go

```diff
@@ -99,14 +107,29 @@ func (s *server) FetchBlob(ctx context.Context, req *pb.FetchBlobRequest) (*pb.F
 	if ar, err := s.storageClient.CheckActionCache(ctx, dg.ToProto()); err != nil {
 		log.Error("Failed to check action cache: %s", err)
 	} else if ar != nil {
-		if len(ar.OutputFiles) == 1 {
-			log.Info("Retrieved %s from action cache (as %s/%d)", req.Uris, dg.Hash, dg.Size)
+		if len(ar.OutputFiles) != 1 {
+			log.Warning("Found %s in action cache (as %s/%d) but it has %d outputs", req.Uris, dg.Hash, dg.Size, len(ar.OutputFiles))
+		} else if s.forceCasCheck {
+			if resp, err := s.storageClient.FindMissingBlobs(ctx, &rpb.FindMissingBlobsRequest{
+				InstanceName: s.storageClient.InstanceName,
+				BlobDigests:  []*rpb.Digest{dg.ToProto()},
+			}); err == nil || len(resp.MissingBlobDigests) == 0 {
+				log.Info("Retrieved %s from action cache (as %s/%d) and exists in CAS", req.Uris, dg.Hash, dg.Size)
+				return &pb.FetchBlobResponse{
+					Status:     &rpcstatus.Status{},
+					BlobDigest: ar.OutputFiles[0].Digest,
+				}, nil
+			} else {
+				//  Missed the CAS, note it down, and move on to usual times.
+				casMissing.Inc()
+			}
+		} else {
+			log.Info("Retrieved %s from action cache (as %s/%d). Did not check CAS", req.Uris, dg.Hash, dg.Size)
```

Comment:
I'm not particularly happy with this spaghetti, but my brain is currently unable to process more complex thoughts.